Inkuire docs (#2125)
scala/docs.scala-lang: scala3/scaladoc.md

![](../resources/images/scala3/scaladoc/snippet-compiler2.gif)
![](../resources/images/scala3/scaladoc/snippet-compiler1.gif)

> ### Type-based searches

Reviewer: ```suggestion
### Type-based search
```

Searching for functions by their symbolic names can be time-consuming.

> That is why the new scaladoc allows to search for functions and fields by their types.

Reviewer: ```suggestion
That is why the new scaladoc allows you to search for methods and fields by their types.
```

![](../resources/images/scala3/scaladoc/snippet-compiler2.gif)
![](../resources/images/scala3/scaladoc/snippet-compiler1.gif)

> ### Inkuire (Hoogle-like searches)

Author: ok, changed

Haskell programmers are probably familiar with [Hoogle](https://hoogle.haskell.org/) - a documentation search engine that allows you to find functions by their signatures rather than their symbolic names. Since many Scala developers are also functional programming fans, we decided to add a similar functionality to Scaladoc.

> To use this feature simply type the signature of the function You are looking for in scaladoc searchbar. This is how it works:

Reviewer: ```suggestion
To use this feature simply type the signature of the function you are looking for in the scaladoc searchbar. This is how it works:
```

Author: done

![](../resources/images/scala3/scaladoc/inkuire-1.0.0-M2_js_flatMap.gif)

> This feature is provided by [Inkuire](https://github.com/VirtusLab/Inkuire) search engine, which works for Scala 3 and Kotlin. To be up-to-date with development of this feature follow [Inkuire repository](https://github.com/VirtusLab/Inkuire) for new releases.

Reviewer: ```suggestion
This feature is provided by the [Inkuire](https://github.com/VirtusLab/Inkuire) search engine, which works for Scala 3 and Kotlin. To be up-to-date with the development of this feature, follow the [Inkuire repository](https://github.com/VirtusLab/Inkuire).
```

Author: done

Note that this feature is still in development, so it can be subject to considerable change.

> If You encounter a bug or have an idea for improvement don't hesitate to create an issue on [Inkuire](https://github.com/VirtusLab/Inkuire/issues/new) or [dotty](https://github.com/lampepfl/dotty/issues/new).

Reviewer: ```suggestion
If you encounter a bug or have an idea for improvement, don't hesitate to create an issue on [Inkuire](https://github.com/VirtusLab/Inkuire/issues/new) or [dotty](https://github.com/lampepfl/dotty/issues/new).
```

Author: done

> Using this feature allows to find `def`s and `val`s from the documented code.

Reviewer: ```suggestion
Using this feature allows you to find `def`s and `val`s from the documented code.
```

Author: done

> In order for a scaladoc searchbar query to be searched using Inkuire instead of the default search engine, the query has to contain `=>` character sequence. 

Reviewer: ```suggestion
In order for a scaladoc searchbar query to be searched using Inkuire instead of the default search engine, the query has to contain the `=>` character sequence. 
```

Author: done

Accepted input is similar to a curried function signature in Scala 3. With some differences:
- AndTypes, OrTypes and Functions have to be enclosed in parentheses e.g. `(Int & Any) => String`

> - Transformations without any input can be found by a syntactic by name call e.g. `=> Int`

Reviewer 2: I think naming those transformations (vals, functions that does take parameters etc.) will make it easier to understand

Author: ok

- A wildcard `_` can be used to indicate that we want to match any type in a given place e.g. `Long => Double => _`

> - Types in the form of single letter e.g. `A` or a letter with a digit `X1` are automatically assumed to be type variables

Reviewer 2: I think it should state here how to look for a type `A` rather then a type variable

Author: Hmmm, it's not possible ATM. Created an issue https://github.com/VirtusLab/Inkuire/issues/199

> - Other type variables can be declared just like in polymorphic functions e.g. `[A, B] => A => B => A` 

Reviewer 2: Shouldn't this example use a name that does not match point above like `Foo`?

Author: Makes sense, done

> - Other type variables can be declared just like in polymorphic functions e.g. `[A, B] => A => B => A` 

Reviewer: So, is this equivalent to?

~~~ scala
A => B => A
~~~

Author: Yes, it is.
As @romanowski suggested, I changed it to `[AVariable, AlsoAVariable] => AVariable => AlsoAVariable => AVariable`.

> Some examples to give You a better idea:

Reviewer: ```suggestion
Some examples to give you a better idea:
```

Author: done

- `List[Int] => (Int => Long) => List[Long]`
- `(A, B) => A`
- `Set[Long] => Long => Boolean`
- `Int => Long => Int`
- `String => Int => Char`

> - `(Int & Float) => (String | Double)`

Reviewer 2: I think we should have few more that uses type variables. I think we may also add some example matches in comments.

Author: ok

Inkuire works as a JavaScript worker in the browser thanks to the power of [ScalaJS](https://www.scala-js.org/).

> To enable Inkuire when running scaladoc add flag `-Ygenerate-inkuire`. By adding this flag two files are generated:

Reviewer: ```suggestion
To enable Inkuire when running scaladoc, add the flag `-Ygenerate-inkuire`. By adding this flag two files are generated:
```

Author: done

> - `inkuire-db.json` - this is the file containing all searchable declarations from the currently documented project in a format readable to Inkuire search engine.

Reviewer: ```suggestion
- `inkuire-db.json` - this is the file containing all the searchable declarations from the currently documented project in a format readable to the Inkuire search engine.
```

Author: done

> - `inkuire-config.json` - this file contains locations of database files that should be searchable from documentation of the current project. By default it will be generated with the location of the local db file as well as default implied locations of detabase files in `External mappings`.

Reviewer 2: We should link to documentation for extenral mapping in scaladoc docs

Author: done

> - `inkuire-config.json` - this file contains locations of database files that should be searchable from documentation of the current project. By default it will be generated with the location of the local db file as well as default implied locations of detabase files in `External mappings`.

Reviewer: ```suggestion
- `inkuire-config.json` - this file contains the locations of the database files that should be searchable from the documentation of the current project. By default, it will be generated with the location of the local db file as well as the default implied locations of database files in `External mappings`.
```

Author: done

> When it comes to how the code is mapped to InkuireDb entries, there are some transformations to make the engine more opinionated(though open to suggestions and changes). Firstly the receiver(non-module owner) of a function can be treated as a first argument. Automatic currying is also applied, so that the results aren't dependant on argument lists. When finding matches `val`s and `def`s are not distinguished.

Reviewer 2: `opinionated(though` and `receiver(non-module owner)` (missing spaces)

Author: done

> When it comes to how the code is mapped to InkuireDb entries, there are some transformations to make the engine more opinionated(though open to suggestions and changes). Firstly the receiver(non-module owner) of a function can be treated as a first argument. Automatic currying is also applied, so that the results aren't dependant on argument lists. When finding matches `val`s and `def`s are not distinguished.

Reviewer: ```suggestion
When it comes to how the code is mapped to InkuireDb entries, there are some transformations to make the engine more opinionated (though open to suggestions and changes). Firstly, the receiver (non-module owner) of a function can be treated as a first argument. Automatic currying is also applied, so that the results don’t depend on argument lists. When finding matches, `val`s and `def`s are not distinguished.
```

Author: done

So the following declarations should be found by query `Int => Int => Int => Int`:
```
class Num:
  def a(i: Int, j: Int): Int
  def b(i: Int)(j: Int): Int
  def c(i: Int): (Int => Int)
  val d: Int => Int => Int
  val e: Int => Int => Int
  val f: (Int, Int) => Int
end Num

def g(i: Num, j: Int, k: Int): Int
extension (i: Num) def h(j: Int, k: Int): Int
def i(i: Num, j: Int)(k: Int): Int
extension (i: Num) def j(j: Int)(k: Int): Int
...

> ```

Reviewer: There is an extra line of:

~~~
```
~~~

Also, it is not clear to me which of the definitions would be found? Actually, I would expect none of them to be found, since all the members of `Num` take two parameters, and the query has three parameters. The methods `g`, `h`, `i`, and `j` include the type `Num`, which is absent from the query.

Author: Oooo, right. I forgot to change the first query argument to `Num` :sweat_smile:
Fixed

Author: That's not an extra line. The previous line is an ellipsis

When it comes to type aliases, they are desugared on both the declaration and the query signature. This means that for declarations:
```
type Name = String

def fromName(name: Name): String
def fromString(str: String): Name
```

> both functions (`fromName` and `fromString`) should be found for queries `Name => Name`, `String => String`, `Name => String` and `String => Name`.

Reviewer: ```suggestion
both motheds, `fromName` and `fromString`, should be found for queries `Name => Name`, `String => String`, `Name => String` and `String => Name`.
```

Author: done

> The edges are the roughest with implicits/givens.

Reviewer 2: I am no sure if we need this sentence

Author: ok, removed
